sdk/nodejs/provider/experimental/analyzer.ts

```diff
@@ -522,6 +537,123 @@ export class Analyzer {
 
         return parts.join(" ");
     }
+
+    /**
+     * Gets the Pulumi resource type information for a resource reference.
+     * A strong assumption is that the referenced resource class is in a package installed to node_modules
+     * and contains a standard Pulumi-generated SDK compiled into JavaScript. To find the resource type token,
+     * the function will attempt to find the JavaScript module file that contains the resource class, and then
+     * extract the type from the __pulumiType property of the resource class. To find the package version,
+     * the function will attempt to read the package.json file in the root directory of the referenced package.
+     * @returns Object containing packageName, packageVersion, and pulumiType token
+     * @throws Error if the resource type cannot be determined with detailed context information
+     */
+    private getResourceType(
+        context: { component: string; property: string; inputOutput: InputOutput; typeName?: string },
+        type: typescript.Type,
+    ): {
+        packageName: string;
+        packageVersion: string;
+        pulumiType: string;
+    } {
+        const symbol = type.getSymbol();
+        if (!symbol) {
+            throw new Error(
+                `Cannot determine resource type: symbol not found for type '${this.checker.typeToString(type)}' for ${this.formatErrorContext(context)}`,
+            );
+        }
+
+        // Try to find the declaration of the class
+        const declaration = symbol.declarations?.[0];
+        if (!declaration) {
+            throw new Error(
+                `Cannot determine resource type: no declarations found for symbol '${symbol.name}' for ${this.formatErrorContext(context)}`,
```

Comment:
These two explanations feel a little too technical maybe?  I'm not a frequent typescript user, but "symbol not found" and "no declarations found" is a bit confusing.  Personally I'd just wrap them all in "could not determine source file for declaration" below or something similar. 

It's a matter of taste though, so feel free to ignore this.